mne_bids/inspect.py

```diff
@@ -205,11 +267,53 @@ def _save_bads(*, bads, bids_path, verbose):
         else:
             # Channel has been manually marked as bad during inspection, assign
             # default description.
-            description = 'Manual inspection via MNE-BIDS'
+            description = 'Interactive inspection via MNE-BIDS'
 
         descriptions.append(description)
 
     # We pass overwrite=True, causing all channels not passed as bad here to
     # be marked as good.
     mark_bad_channels(ch_names=bads, descriptions=descriptions,
                       bids_path=bids_path, overwrite=True, verbose=verbose)
+
+
+def _save_annotations(*, annotations, bids_path, verbose):
+    # Read the raw data, set our new Annotations, and convert them to events
+    # which we can then stored in the *_events.tsv sidecar.
+    extra_params = dict(preload=True)
+    if bids_path.extension == '.fif':
+        extra_params['allow_maxshield'] = True
+
+    raw = read_raw_bids(bids_path=bids_path, extra_params=extra_params,
+                        verbose='warning')
+    raw.set_annotations(annotations)
+
+    events, durs, descrs = _read_events(events_data=None, event_id=None,
+                                        raw=raw, verbose=False)
+
+    # Write raw data.
+    if isinstance(raw, mne.io.brainvision.brainvision.RawBrainVision):
+        # XXX We should write durations too, this is supported by pybv.
+        _write_raw_brainvision(raw, bids_path.fpath, events)
+    elif isinstance(raw, mne.io.RawFIF):
+        raw.save(raw.filenames[0], overwrite=True, split_naming='bids',
+                 verbose=False)
```

Comment:
to me we should update the side car files and not alter the raw fif files

Reply:
cf discussion with Robert

Reply:
Yes this would make easier.
Just doing it this way here because that's also how we're currently handling bad channel marking as well – was trying to be consistent. But I will gladly remove this…
@sappelhoff WDYT?

Reply:
Removing this would also GREATLY speed up operation

Reply:
just to clarify: would that mean that when people just load the FIF and ignore the bids events.tsv, they get the "poor, non-updated data"?

I would be fine with this. It's also how I handle my own (BIDS) data --> my events.tsv contains significantly more information than the raw files, and I don't sync them both every time I add an annotation or similar.

However, I *do* make sure that events.tsv is a superset of what's available in the raw data. :thinking: ... not sure if that's necessary though. Software could handle merging the two sets of reading (i) BIDS sidecar, and (ii) raw data

Reply:
> Software could handle merging the two sets of reading (i) BIDS sidecar, and (ii) raw data

This is giving me the chills just thinking about it :)

My proposal would be:

- our write functions should drop remove some the from the raw file that can be repopulated from the sidecars: bads, events (annotations), subject info
- sidecars become the single source of truth

Reply:
I've addressed this in bafb6f2

Reply:
BIDS sidecar files have precedence. They are the "truth". To me if they are present they define what are the annotations. If you don't use a BIDS file reader for your files then too bad. You have what is in the manufacturer file as written during acquisition. That's the vision to me. See also discussion with Robert O. recently. He said the same thing.

Reply:
> If you don't use a BIDS file reader for your files then too bad.

Yes I totally agree. It would be like complaining that you don't get proper formatting of a modern website if you view it in links on your Linux terminal – you're not using the tool the website was designed for. I guess the unfortunate thing is that users are still able to **open** the files, instead of being told, "hey, unsupported format, use xxx instead"

Reply:
> You have what is in the manufacturer file as written during acquisition. That's the vision to me.

Ok I can live with that. Although I'm still concerned that users might e.g. assume a file is anonymized if they simply adjust the sidecars.

Reply:
Still – that's **our** opinion. It's not what the specs say. We need to push forward an amendment of the BIDS specification.

Reply:
yes, that's probably a good idea.

> our write functions (write_raw_bids, write_anat) should drop all info from the raw file

as long as we do not lose resolution ... example: dropping the BrainVision annotations from vmrk --> these are in samples (1-indexed, I think) ... that means in BIDS sidecar events.tsv there MUST be a column `sample`, which can relate a given event back to the exact sample in the data --> not something in seconds.

also, we cannot drop all data unfortunately --> some files will require things like dates, etc. 

Other than that +1